jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java

```diff
@@ -801,9 +826,10 @@ public void write(byte[] b, int off, int len) throws IOException
                 channelWrite(_aggregate, last && len == 0);
 
                 // should we fill aggregate again from the buffer?
-                if (len > 0 && !last && len <= _commitSize && len <= BufferUtil.space(_aggregate))
+                if (len > 0 && !last && len <= _commitSize && len <= getAggregateSpace())
                 {
                     BufferUtil.append(_aggregate, b, off, len);
+                    onWriteComplete(false, null);
                     return;
                 }
```

Comment:
This code block should really be removed, so that aggregation will happen only in 1 place, and only due to computing whether it should happen or not only once.

The conditions at line 829 (new code) are slightly different from the conditions at line 757 (new code), so it's difficult to follow the code, trying to understand why in the world there is a call to `onWriteComplete()` inside a block that does aggregation.

Actually, turns out that we do (again) a similar aggregation logic in `AsyncWrite.process()`, so we need to clean this up even more.

Reply:
I tend to agree, but for jetty-10 only.  In Jetty-9 I do not think we should be making significant behaviour changes as they tend to have unintended consequences.

Reply:
I am also concerned about the performance. Consider an app that always writes small 5, 10, 20 byte writes. Eventually a small write overflows the buffer. Currently the code fills the buffer to max size, flushes it and then aggregates the small remaining data.
The alternative is to flush a non full buffer and then write a small write, so we get 2 non optimal writes instead of one.

The real solution is to add gather writes in jetty-10 or beyond